<center><p float="center">
  <img src="https://upload.wikimedia.org/wikipedia/commons/e/e9/4_RGB_McCombs_School_Brand_Branded.png" width="300"/>
  <img src="https://mma.prnewswire.com/media/1458111/Great_Learning_Logo.jpg?p=facebook" width="200"/>
</p></center>

<center><font size=10>AI Agents and Generative AI for Business Applications</center></font>
<center><font size=6>Retrieval Augmented Generation</center></font>

<center><p float="center">
  <img src="https://images.pexels.com/photos/8441780/pexels-photo-8441780.jpeg" width="640"/>
</p></center>

<center><font size=6>Health Insurance Q&A</center></font>

# Problem Statement

### Business Context

MeridianHealth Insurance is a mid-sized health insurance carrier headquartered in Hartford, Connecticut, licensed to operate across 22 states with concentrated presence in the Northeast, Mid-Atlantic, and Midwest. The company serves approximately 2.4 million covered lives across a portfolio of individual, family, employer-sponsored, and supplemental health insurance products.

MeridianHealth booked roughly $4.1 billion in earned premiums last fiscal year and operates through a hybrid distribution model, a direct sales force of close to 4,800 licensed sales representatives, a network of 28,000 appointed independent agents and brokers, and 14 regional employer-group partnerships. The product portfolio currently spans 47 active plan designs, each with its own combination of deductibles, coinsurance, copay structures, out-of-pocket maximums, waiting periods, claim rules, exclusions, wellness benefits, rider options, and in-network provider arrangements that vary by state and rating area.

The core business process at the centre of this case is the live sales conversation, the moment a sales representative is on a call, a video meeting, or sitting across the table from a prospective customer, trying to convert interest into an enrolled member. During these conversations, customers ask sharp, specific questions, such as:

- Whether maternity is covered under a specific family floater plan, and what waiting period applies
- What the deductible and co-payment look like under a high-deductible health plan
- Whether a pre-existing condition, like diabetes or hypertension, is covered from day one
- Whether wellness rewards, cumulative bonuses, or recharge benefits continue after renewal
- Whether a particular expense, consumable, or treatment falls under the claim exclusion list
- What documents are required for reimbursement or cashless hospitalisation claims

The sales representative is expected to answer these accurately and on the spot while staying compliant with disclosure and policy wording requirements. The information needed to answer these questions is scattered across long policy brochures, policy terms and conditions, claim exclusion documents, reimbursement and cashless claim guides, FAQs, underwriting circulars, and product training material running into thousands of pages across the active portfolio.

The problem is that sales representatives cannot reliably retrieve this information in real time during a live conversation. New representatives, who form roughly 35% of the field force at any given time given an industry attrition rate of 28–30%, often do not know where the answer lives, let alone what it says. Even experienced representatives struggle when:

- The question requires combining information from multiple policy documents
- The clause involves waiting periods, exclusions, or conditional coverage rules
- The answer depends on interpreting both policy wording and claim guidance together
- The answer is buried deep inside a long Terms & Conditions or policy brochure PDF

The most common workarounds calling a back-office product helpdesk, putting the customer on hold, or making a confident guess, each carry a cost. The product helpdesk receives an estimated 11,000–13,000 internal queries per month from the field, with average resolution times of 6–14 minutes per query. When the representative guesses or paraphrases incorrectly, the risk shifts to policy misrepresentation, claim disputes, customer dissatisfaction, and regulatory scrutiny.

The operational and human consequences are significant:
- An estimated 18–22% of sales conversations stall or drop off at the question-handling stage, with internal win-loss reviews indicating “couldn’t answer customer’s clarification” as a recurring loss reason.
- Misrepresentation complaints, where the customer alleges they were told a benefit existed that the policy does not actually provide contributed to substantial remediation costs, claim escalations, and compliance investigations last year.
- New sales representatives take 9–12 months to reach full productivity, largely because product mastery across dozens of policy variations is built through experience rather than accessible knowledge retrieval.
- Senior product specialists and operations teams spend a large portion of their working hours answering repetitive field queries, reducing time available for underwriting support, compliance review, and product enablement.
- The sales representative on the front line carries the cognitive and emotional load of being asked questions they cannot confidently answer in front of a customer who is deciding whether to trust them with their family’s healthcare coverage.

### Objective

MeridianHealth is initiating a proof-of-concept to build a Retrieval-Augmented Generation (RAG)-based policy intelligence system that allows a sales representative to retrieve accurate, source-grounded answers from across the company’s health insurance documentation in real time, while also rigorously evaluating answer quality before deployment in a regulated environment.

- The system should ingest policy brochures, policy terms and conditions, claim exclusion references, and claim process guides into a searchable vector database so that relevant clauses and sections can be retrieved semantically during a live customer interaction.
- The solution should read a sales representative’s question, typed or spoken, retrieve the most relevant policy context, and generate a precise, grounded answer using a Large Language Model, enabling the representative to respond within the flow of the conversation instead of putting the customer on hold or escalating to the back office.
- The system should return the supporting source passages and document references alongside every generated answer so that representatives can verify the wording, explain the reasoning confidently, and maintain a transparent audit trail of what information was surfaced during the conversation.
- The solution should also handle comparative and cross-document questions, such as checking differences between policy benefits and exclusions, validating waiting-period conditions, or combining reimbursement and cashless claim guidance, so that representatives can guide customers accurately across multiple scenarios instead of relying on memory or partial knowledge.
- To improve reliability, the system should include a prompt optimization and evaluation layer using DeepEval and GEPA-based prompt tuning. The evaluation framework should continuously measure answer relevance, faithfulness to retrieved context, and overall response quality across a benchmark dataset of realistic sales and policy questions.
- The proof-of-concept should further evaluate multiple runtime RAG configurations, including retrieval depth and generation parameters, to identify the most reliable setup for grounded policy answering in production-like scenarios.

A successful pilot, measured through strong benchmark performance, accurate source-grounded responses, reduced dependency on the back-office helpdesk, and improved confidence during customer conversations, will give MeridianHealth the confidence to scale the capability across its direct sales force and later to its broker and partner ecosystem, while reducing misrepresentation risk and shortening time-to-productivity for new representatives.


### Data Description

The system uses four health insurance documents stored locally for the RAG workflow. These files together cover product details, policy wording, exclusions, and claims handling. The files are: `Policy_Brocher.pdf`, `Health_Insurance_Policy_TandC.pdf`, `List_of_Standard_Claim_Exclusions.pdf`, and `Health_Insurance_Claim_Guide.pdf`.
1. **Policy Brochure** `(Policy_Brocher.pdf)`:
 > Contains the main product overview in simple language, including family floater plan details, benefits, coverage highlights, exclusions, eligibility, how the plan works, buying steps, claims, renewals, member addition rules, and premium factors.

2. **Health Insurance Policy Terms & Conditions** `(Health_Insurance_Policy_TandC.pdf)`:
> Contains the formal policy wording, including definitions, benefits covered, general conditions, waiting periods, policy rules, claim conditions, renewal terms, and other legal policy clauses used for grounded answers.

3. **List of Standard Claim Exclusions** `(List_of_Standard_Claim_Exclusions.pdf)`:
> Contains items and charges that are not payable under a standard health claim, along with items that are only payable under specific policy wording. It also lists hospital-service and room-charge items that cannot be billed separately.

4. **Health Insurance Claim Guide** `(Health_Insurance_Claim_Guide.pdf)`:
> Contains the reimbursement and cashless claim process, document requirements, claim status steps, and common FAQ guidance for policyholders.


# Installing and Importing Necessary Libraries and Dependencies

In [1]:
# Installing the necessary libraries with specified versions
%pip install -qU \
    "chromadb==1.5.9" \
    "langchain-community==0.4.1" \
    "langchain-chroma==1.1.0" \
    "langchain-openai==1.2.1" \
    "langchain-text-splitters==1.1.2" \
    "pandas>=2.2.0" \
    "numpy>=1.26.0" \
    "scikit-learn>=1.4.0" \
    "python-dotenv>=1.0.1" \
    "tqdm>=4.66.0" \
    "pypdf>=5.0.0" \
    "deepeval"


[notice] A new release of pip is available: 26.1.2 -> 26.2.1
[notice] To update, run: pip3 install --upgrade pip
Note: you may need to restart the kernel to use updated packages.


**Note**:
- After running the above cell, kindly restart the runtime (for Google Colab) or notebook kernel (for VSCode), and run all cells sequentially from the next cell.
- On executing the above line of code, you might see a warning regarding package dependencies. This error message can be ignored as the above code ensures that all necessary libraries and their dependencies are maintained to successfully execute the code in ***this notebook***.

***Prompt:***

<font size=3 color="#4682B4"><b>I want to build and evaluate a Retrieval-Augmented Generation (RAG) system for health insurance policy intelligence. Help me install and import the necessary Python libraries to:

1. Read and manipulate structured data such as JSON and tabular datasets
2. Work with environment variables, file paths, and runtime utilities
3. Load and process health insurance policy PDF documents
4. Split long policy documents into smaller semantic chunks
5. Generate vector embeddings using OpenAI embedding models
6. Store and retrieve embeddings using Chroma vector database
7. Build RAG pipelines using LangChain components and chat models
8. Optimize prompts using DeepEval and the GEPA prompt optimization algorithm
9. Evaluate generated answers using relevance, faithfulness, and completeness metrics
10. Create and manage benchmark datasets for prompt optimization and RAG evaluation

</font>

<font size=3 color="#4682B4"><b>

</font>

In [2]:
# Importing the necessary libraries

# Standard Python libraries for file handling, text processing, JSON handling,
# randomness, timing, and basic utilities.
import os
import re
import json
import random
import time
from pathlib import Path
from collections import Counter
import pandas as pd

# DeepEval libraries for prompt optimization and evaluation.
from deepeval.prompt import Prompt
from deepeval.dataset import Golden
from deepeval.metrics import GEval
from deepeval.optimizer import PromptOptimizer
from deepeval.optimizer.algorithms import GEPA
from deepeval.test_case import LLMTestCase, SingleTurnParams
from deepeval.models import GPTModel
from deepeval.optimizer.policies import TieBreaker

# LangChain libraries for document loading, text splitting, embeddings,
# vector storage, and LLM interaction.
from langchain_community.document_loaders import PyPDFLoader
from langchain_text_splitters import RecursiveCharacterTextSplitter
from langchain_chroma import Chroma
from langchain_openai import OpenAIEmbeddings, ChatOpenAI
from langchain_core.messages import SystemMessage, HumanMessage

/Library/Frameworks/Python.framework/Versions/3.14/lib/python3.14/site-packages/tqdm/auto.py:21: TqdmWarning: IProgress not found. Please update jupyter and ipywidgets. See https://ipywidgets.readthedocs.io/en/stable/user_install.html
  from .autonotebook import tqdm as notebook_tqdm
/var/folders/rl/rvl8m81x0ydgy38wjpr83s9m0000gn/T/ipykernel_66532/501686957.py:21: DeprecationWarning: 'GPTModel' is deprecated and will be removed in a future release. Use 'OpenAIModel' instead.
  from deepeval.models import GPTModel


# Environment Setup and Project Initialization

***Prompt:***

<font size=3 color="#4682B4"><b>I want to initialize the environment for a health insurance policy intelligence RAG project. Help me:

1. Define local file paths for all health insurance policy PDF documents and the benchmark evaluation dataset
2. Create local directories for Chroma vector database persistence and evaluation artifacts
3. Automatically create directories if they do not already exist
4. Configure OpenAI API credentials using environment variables
5. Define the embedding model and answer-generation model used throughout the notebook
6. Display the active project directories for verification

</font>

<font size=3 color="#4682B4"><b>

</font>

In [ ]:
# Define the local paths for all input datasets.
# All PDFs and local project files are stored directly in the project folder.

BASE_DIR = Path("/Users/kavitha/Downloads/sample RAG project")
DATA_DIR = BASE_DIR

POLICY_BROCHURE_PDF = DATA_DIR / "Policy_Brocher.pdf"
POLICY_TERMS_PDF = DATA_DIR / "Health_Insurance_Policy_TandC.pdf"
CLAIM_EXCLUSIONS_PDF = DATA_DIR / "List_of_Standard_Claim_Exclusions.pdf"
CLAIM_GUIDE_PDF = DATA_DIR / "Health_Insurance_Claim_Guide.pdf"

BENCHMARK_CSV = DATA_DIR / "golden_benchmark_dataset.csv"

In [ ]:
# Define local directories for:
# 1. Chroma vector database persistence
# 2. Saved DeepEval optimized prompts, evaluation outputs, and RAG tuning artifacts

CHROMA_DIR = DATA_DIR / "chroma_db"
ARTIFACT_DIR = DATA_DIR / "artifacts"

# Create the directories automatically if they do not already exist.
CHROMA_DIR.mkdir(exist_ok=True, parents=True)
ARTIFACT_DIR.mkdir(exist_ok=True, parents=True)

In [5]:
# Load the credentials JSON file and extract values
# These credentials are used for:
# - embedding generation
# - answer generation
# - DeepEval optimization and evaluation
file_name = 'config.json'                                                       # Name of the configuration file
with open(file_name, 'r') as file:                                              # Open the config file in read mode
    config = json.load(file)                                                    # Load the JSON content as a dictionary
    OPENAI_API_KEY = config.get("OPENAI_API_KEY")                                             # Extract the API key from the config
    OPENAI_API_BASE = config.get("OPENAI_API_BASE")

# Store API credentials in environment variables
os.environ['OPENAI_API_KEY'] = OPENAI_API_KEY                                  # Set API key as environment variable
os.environ["OPENAI_BASE_URL"] = OPENAI_API_BASE                                 # Set API base URL as environment variable

In [6]:
# Define the models used throughout the notebook.

# Embedding model:
# Converts health insurance policy text into vector embeddings.
EMBEDDING_MODEL = "text-embedding-3-small"

# Answer generation model:
# Used for baseline RAG generation, DeepEval prompt optimization,
# RAG configuration tuning, and final policy intelligence inference.
ANSWER_MODEL = "gpt-4o-mini"
EVAL_MODEL = "gpt-4o-mini"

In [7]:
# Display the active working directories.

print("Data directory:", DATA_DIR.resolve())
print("Chroma directory:", CHROMA_DIR.resolve())
print("Artifacts directory:", ARTIFACT_DIR.resolve())

Data directory: /Users/kavitha/Downloads/sample RAG project/data
Chroma directory: /Users/kavitha/Downloads/sample RAG project/chroma_db
Artifacts directory: /Users/kavitha/Downloads/sample RAG project/artifacts


# Section 1: Baseline Retrieval-Augmented Generation (RAG) Pipeline

## 1.1 Load the PDF files with `PyPDFLoader`

***Prompt:***

<font size=3 color="#4682B4"><b>I want to load multiple health insurance policy PDF documents into LangChain so that each page can be used as a retrievable document. Help me create functions to:

1. Check whether the required PDF file exists locally before loading it
2. Load each PDF using `PyPDFLoader`
3. Convert every page into a LangChain document
4. Add source-level metadata such as file name and file type for traceability
5. Load the policy brochure, policy terms and conditions, claim exclusions, and claim guide as separate document sets

</font>

In [8]:
def ensure_file_exists(path: Path) -> None:
    """
    Check whether the required input file exists locally.

    Raises a clear error if the file is missing so the pipeline
    stops before document loading begins.
    """
    if not path.exists():
        raise FileNotFoundError(
            f"Missing required file: {path}. Put it under {DATA_DIR.resolve()} and run again."
        )

In [9]:
def load_pdf_documents(path: Path):
    """
    Load a PDF file using LangChain's PyPDFLoader.

    Each page becomes one LangChain Document, and the source file
    information is preserved for traceability.
    """
    ensure_file_exists(path)

    loader = PyPDFLoader(str(path))
    docs = loader.load()

    # Add dataset-level metadata to every document
    # so the source type remains identifiable later.
    for doc in docs:
        doc.metadata["source_file"] = path.name
        doc.metadata["source_type"] = "pdf"

    return docs

In [12]:
# Load policy brochure pages as LangChain Documents.
brochure_docs = load_pdf_documents(POLICY_BROCHURE_PDF)

# Load policy terms and conditions pages as LangChain Documents.
terms_docs = load_pdf_documents(POLICY_TERMS_PDF)

# Load standard claim exclusions pages as LangChain Documents.
exclusions_docs = load_pdf_documents(CLAIM_EXCLUSIONS_PDF)

# Load claim guide pages as LangChain Documents.
claim_guide_docs = load_pdf_documents(CLAIM_GUIDE_PDF)

# Display basic verification information
# to confirm document loading worked correctly.
print(f"Loaded brochure documents:   {len(brochure_docs)}")
print(f"Loaded terms documents:      {len(terms_docs)}")
print(f"Loaded exclusions documents: {len(exclusions_docs)}")
print(f"Loaded claim guide docs:     {len(claim_guide_docs)}")

print("Brochure doc example:")
print(brochure_docs[0].page_content[:700])

print("\nMetadata:", brochure_docs[0].metadata)

FileNotFoundError: Missing required file: data/Policy_Brocher.pdf. Put it under /Users/kavitha/Downloads/sample RAG project/data and run again.

## 1.2 Load the PDF pages and split them into chunks



***Prompt:***

<font size=3 color="#4682B4"><b>I want to prepare the loaded PDF pages for semantic retrieval in a RAG pipeline. Help me write code to:

1. Combine all loaded policy documents into one dataset
2. Split the documents into smaller text chunks using `RecursiveCharacterTextSplitter`
3. Use a chunk size and overlap that preserve context across adjacent chunks
4. Create chunked documents that are suitable for embeddings and vector search
5. Print verification details such as total source documents, total chunks, and sample chunk metadata

</font>

In [ ]:
# Combine all source documents before chunking.
all_source_docs = brochure_docs + terms_docs + exclusions_docs + claim_guide_docs

# Configure the text splitter used for chunking.
#   chunk_size: Maximum number of characters per chunk.
#   chunk_overlap: Overlapping text between chunks to preserve context continuity.
splitter = RecursiveCharacterTextSplitter(
    chunk_size=1200,
    chunk_overlap=150,
)

# Split the PDF pages into smaller chunks for embedding generation and semantic retrieval.
policy_chunks = splitter.split_documents(all_source_docs)

# Display verification information to confirm chunking worked correctly.
print(f"Loaded source documents: {len(all_source_docs)}")
print(f"Split policy chunks:     {len(policy_chunks)}")

print("Policy chunk example:")
print(policy_chunks[0].page_content[:700])

print("\nMetadata:", policy_chunks[0].metadata)

Loaded source documents: 70
Split policy chunks:     195
Policy chunk example:
Family Health Insurance Plans 
A complete guide to protecting your loved ones with a single, shared health insurance plan — covering 
benefits, features, eligibility, claims, exclusions, and how to choose the right cover for your household. 
 
 
What Is a Family Health Insurance Plan? 
A family health insurance plan, often called a family floater, allows you to cover several members of your 
household — typically up to two adults and two children — under one single policy. You pay a single 
premium, and everyone insured shares one common sum insured. Instead of juggling multiple individual 
policies, your family gets one consolidated cover that any insured member can draw on during a medical

Metadata: {'producer': 'www.ilovepdf.com', 'creator': 'Microsoft® Word 2016', 'creationdate': '2026-05-07T09:55:46+00:00', 'moddate': '2026-05-07T09:55:47+00:00', 'source': 'data/Policy_Brocher.pdf', 'total_pages': 31, 

## 1.3 Build a local Chroma vector store

***Prompt:***

<font size=3 color="#4682B4"><b>I want to store the chunked health insurance policy documents in a local vector database for retrieval. Help me write code to:

1. Generate embeddings for the chunked documents using OpenAI embeddings
2. Create a local Chroma vector store
3. Persist the vector database in a local directory
4. Store all chunked policy documents in one searchable collection
5. Print confirmation that the vector store is ready and the documents have been indexed

</font>

In [ ]:
# Create embeddings and index everything into Chroma.
embeddings = OpenAIEmbeddings(model=EMBEDDING_MODEL)

all_documents = policy_chunks

# Chroma persists locally inside CHROMA_DIR.
vectorstore = Chroma.from_documents(
    documents=all_documents,
    embedding=embeddings,
    collection_name="health_policy_intelligence",
    persist_directory=str(CHROMA_DIR),
)

print(f"Indexed documents: {len(all_documents)}")
print("Vector store is ready.")

Indexed documents: 195
Vector store is ready.


## 1.4 Retrieve Context and Generate Answers using the Baseline RAG Pipeline


***Prompt:***

<font size=3 color="#4682B4"><b>I want to build a simple baseline RAG pipeline for answering health insurance policy questions. Help me write code to:

1. Initialize a chat model for answer generation
2. Retrieve the most relevant chunks from the Chroma vector store for a user question
3. Format retrieved documents into a compact context block with metadata
4. Create a very simple baseline system prompt and user prompt
5. Keep the question and retrieved context dynamic during runtime
6. Generate an answer using only the retrieved policy context
7. Return the answer, retrieved documents, and formatted context from a reusable function
8. Run one sample policy question to verify the full end-to-end workflow

</font>

In [ ]:
# Initialize the answer-generation language model.
# temperature=0 keeps responses more deterministic and consistent.
llm = ChatOpenAI(
    model=ANSWER_MODEL,
    temperature=0,
    api_key=os.getenv("OPENAI_API_KEY"),
    base_url=os.getenv("OPENAI_BASE_URL"),
)

In [ ]:
def retrieve_docs(question: str, k: int = 4):
    """
    Retrieve the most semantically relevant document chunks
    from the Chroma vector database.
    """
    return vectorstore.similarity_search(question, k=k)


def format_docs(docs, max_chars: int = 1200) -> str:
    """
    Convert retrieved documents into a compact context block.

    The formatted context is later inserted into the prompt
    before sending it to the language model.
    """
    blocks = []

    for i, doc in enumerate(docs, start=1):
        # Clean and shorten the retrieved text chunk.
        snippet = doc.page_content.strip().replace("\n", " ")

        if len(snippet) > max_chars:
            snippet = snippet[:max_chars] + "..."

        # Attach metadata so retrieved sources remain visible.
        blocks.append(
            f"[Source {i}] {snippet}\n"
            f"Metadata: {json.dumps(doc.metadata, ensure_ascii=False)}"
        )

    return "\n\n".join(blocks)

In [ ]:
# -------------------------------------------------------------------
# Baseline RAG Prompt Templates
# -------------------------------------------------------------------

BASELINE_SYSTEM_PROMPT = """
You are a health insurance assistant.
Answer the question using the context.
""".strip()

In [ ]:
BASELINE_USER_PROMPT = """
Broker Question:
{question}

Retrieved Context:
{context}
""".strip()

**Note**: The following part of the code in the `BASELINE_USER_PROMPT` ensures that the question and context values are dynamically fetched during runtime.

```
Broker Question:
{question}
Retrieved Context:
{context}
```

Please ensure that this part remains UNCHANGED.

In [ ]:
def baseline_answer(question: str, k: int = 4) -> dict:
    """
    Generate an answer using the baseline RAG workflow.

    Steps:
    1. Retrieve relevant policy context
    2. Build the grounded prompt
    3. Send the prompt to the LLM
    4. Return the generated answer and retrieved sources
    """

    # Retrieve relevant chunks from Chroma
    docs = retrieve_docs(question, k=k)

    # Convert retrieved documents into prompt-ready context
    context = format_docs(docs)

    # Inject the runtime values into the prompt template
    user_prompt = BASELINE_USER_PROMPT.format(
        question=question,
        context=context,
    )

    # Generate the final response
    response = llm.invoke([
        SystemMessage(content=BASELINE_SYSTEM_PROMPT),
        HumanMessage(content=user_prompt),
    ])

    return {
        "answer": response.content,
        "docs": docs,
        "context": context,
    }

In [ ]:
# Run one sample query to verify the end-to-end RAG workflow.
sample_question = (
    "What documents are required for a reimbursement claim, and how does the cashless claim process work?"
)

sample_result = baseline_answer(sample_question)

print(sample_result["answer"])

For a reimbursement claim, the following documents are typically required:

1. A duly completed claim form.
2. Original policy documents.
3. Original receipts from the hospital.
4. Medical bills.
5. Medical reports.
6. Discharge summary.

In terms of the cashless claim process, here are the steps involved:

1. **Locate a network hospital** near you.
2. At the **third-party administrator (TPA) desk** of the hospital, complete the **pre-authorisation request form**.
3. Once the request is received from the hospital, the insurer begins the **approval process** and keeps you updated on the claim status. You can also track the claim through the insurer's mobile app.
4. All further coordination and documentation is handled directly with the hospital's TPA desk.
5. The insurer settles the bill directly with the hospital at the time of discharge.

If the cashless request is rejected, you can still proceed with treatment and later apply for a reimbursement claim.


## 1.5 Load the gold benchmark dataset and evaluate the baseline prompt


***Prompt:***

<font size=3 color="#4682B4"><b>I want to prepare a gold benchmark dataset for evaluating my health insurance RAG system. Help me write code to:

1. Load a CSV benchmark dataset containing question, answer, context, and supporting sources
2. Validate that all required columns are present before using the dataset
3. Convert each row into a structured dictionary format
4. Create DeepEval Golden objects from the benchmark rows
5. Split the benchmark into train and test sets
6. Print the number of examples in each split for verification

</font>

In [ ]:
def load_gold_benchmark(path: Path) -> list[dict]:
    """
    Load the benchmark dataset created for health insurance policy evaluation.

    Required columns:
    - question
    - answer
    - context
    - supporting_sources
    """
    ensure_file_exists(path)

    df = pd.read_csv(path)

    required_cols = ["question", "answer", "context", "supporting_sources"]
    missing_cols = [col for col in required_cols if col not in df.columns]
    if missing_cols:
        raise ValueError(
            f"Missing required benchmark columns: {missing_cols}. "
            f"Available columns: {list(df.columns)}"
        )

    rows = []
    for _, row in df.iterrows():
        rows.append({
            "question": str(row["question"]).strip(),
            "answer": str(row["answer"]).strip(),
            "context": str(row["context"]).strip(),
            "supporting_sources": str(row["supporting_sources"]).strip(),
        })

    return rows


In [ ]:
# Load the benchmark examples.
gold_rows = load_gold_benchmark(BENCHMARK_CSV)

if len(gold_rows) != 30:
    raise ValueError(f"Expected 30 gold examples, but found {len(gold_rows)}.")


# Build DeepEval Golden objects.
goldens = []
for row in gold_rows:
    goldens.append(
        Golden(
            input=row["question"],
            expected_output=row["answer"],
            context=[row["context"]],
        )
    )


# Split into 24 training examples and 6 test examples.
random.seed(42)
random.shuffle(goldens)

train_goldens = goldens[:24]
test_goldens = goldens[24:]


# Display verification information
print(f"Gold examples: {len(goldens)}")
print(f"Trainset:      {len(train_goldens)}")
print(f"Test set:      {len(test_goldens)}")

print("\nSample golden:")
print(train_goldens[0])

## 1.6 Define the evaluation metrics


***Prompt:***

<font size=3 color="#4682B4"><b>I want to evaluate a health insurance RAG system using benchmark answers and retrieved context. Help me define evaluation metrics to measure:

1. Relevance between the generated answer and the expected answer
2. Faithfulness of the generated answer to the retrieved policy context
3. Completeness of the answer in covering the important policy details
4. A DeepEval GPT-based evaluation model to score the outputs
5. A final list of metrics that can be reused across baseline evaluation, prompt optimization, and runtime tuning

</font>

In [ ]:
# Define the evaluation criteria.

RELEVANCE_CRITERIA = (
    "Evaluate whether the answer is factually aligned with the expected answer. "
    "Reward answers that preserve the key policy, claim, coverage, exclusion, "
    "waiting-period, or eligibility details and penalize missing or incorrect claims."
)

FAITHFULNESS_CRITERIA = (
    "Evaluate whether the answer stays grounded in the retrieved policy evidence. "
    "Reward answers that clearly rely on retrieved brochure, policy wording, exclusions, "
    "or claim guidance. Penalize unsupported or hallucinated policy statements."
)

COMPLETENESS_CRITERIA = (
    "Evaluate whether the answer fully addresses the policy question using the available context. "
    "Reward answers that cover all important policy points, exclusions, claim conditions, "
    "or procedural details. Penalize answers that are incomplete or miss key information."
)

In [ ]:
# Evaluation model used by DeepEval metrics.
eval_gpt_model = GPTModel(
    model=EVAL_MODEL,
    api_key=os.getenv("OPENAI_API_KEY"),
    base_url=os.getenv("OPENAI_BASE_URL"),
)

# 1. Relevance:
# Checks whether the generated answer aligns with the expected benchmark answer.
relevance_metric = GEval(
    name="Relevance",
    criteria=RELEVANCE_CRITERIA,
    evaluation_params=[
        SingleTurnParams.ACTUAL_OUTPUT,
        SingleTurnParams.EXPECTED_OUTPUT,
    ],
    model=eval_gpt_model,
    threshold=0.7,
)

# 2. Faithfulness:
# Checks whether the answer stays grounded in the retrieved context.
faithfulness_metric = GEval(
    name="Faithfulness",
    criteria=FAITHFULNESS_CRITERIA,
    evaluation_params=[
        SingleTurnParams.ACTUAL_OUTPUT,
        SingleTurnParams.CONTEXT,
    ],
    model=eval_gpt_model,
    threshold=0.7,
)

# 3. Completeness:
# Checks whether the answer fully covers the important aspects of the question.
completeness_metric = GEval(
    name="Completeness",
    criteria=COMPLETENESS_CRITERIA,
    evaluation_params=[
        SingleTurnParams.INPUT,
        SingleTurnParams.ACTUAL_OUTPUT,
        SingleTurnParams.CONTEXT,
    ],
    model=eval_gpt_model,
    threshold=0.7,
)

# Final metric list used across:
# - baseline evaluation
# - GEPA optimization
# - RAG configuration tuning
DEEPEVAL_METRICS = [
    relevance_metric,
    faithfulness_metric,
    completeness_metric,
]


## 1.7 Evaluate the baseline RAG on the golden examples dataset

***Prompt:***

<font size=3 color="#4682B4"><b>I want to evaluate my baseline health insurance RAG pipeline against a benchmark dataset. Help me write code to:

1. Create a reusable evaluation function for any answer-generation pipeline
2. Run the pipeline on each golden example question
3. Compare the generated answer with the expected answer and retrieved context
4. Measure relevance, faithfulness, and completeness using DeepEval
5. Collect metric scores across all examples
6. Compute average metric values and an overall score
7. Display the final evaluation summary for the baseline RAG pipeline

</font>

In [ ]:
def evaluate_pipeline_on_goldens(
    answer_function,
    goldens_subset,
    label: str,
) -> dict:
    """
    Evaluate any RAG pipeline on a subset of benchmark gold examples.

    Parameters:
    - answer_function:
        A function that takes a question and returns:
        {
            "answer": ...,
            "context": ...
        }

    - goldens_subset:
        Train or test goldens

    - label:
        Display label for logs/results
    """

    metric_scores = {
        metric.name: []
        for metric in DEEPEVAL_METRICS
    }

    for idx, golden in enumerate(goldens_subset, start=1):

        print(f"\nEvaluating Example {idx}/{len(goldens_subset)} [{label}]")

        # Generate pipeline output
        result = answer_function(golden.input)

        # Build DeepEval test case
        test_case = LLMTestCase(
            input=golden.input,
            actual_output=result["answer"],
            expected_output=golden.expected_output,
            context=[result["context"]],
        )

        # Run evaluation metrics
        for metric in DEEPEVAL_METRICS:
            try:
                metric.measure(test_case)

                if metric.score is not None:
                    metric_scores[metric.name].append(metric.score)

                print(f"{metric.name}: {metric.score}")

            except Exception as e:
                print(f"{metric.name} failed on this example: {e}")

    # Compute metric means
    metric_means = {
        name: round(sum(scores) / len(scores), 3)
        if scores else 0.0
        for name, scores in metric_scores.items()
    }

    # Compute overall score
    overall_mean = round(
        sum(metric_means.values()) / len(metric_means),
        3
    ) if metric_means else 0.0

    print("\n" + "=" * 80)
    print(f"{label} Metric Means: {metric_means}")
    print(f"{label}: Final Mean DeepEval Score = {overall_mean}")
    print("=" * 80)

    return {
        "label": label,
        "metric_means": metric_means,
        "overall_mean": overall_mean,
    }

In [ ]:
# -------------------------------------------------------------------
# Evaluate Baseline RAG only on the TEST set
# -------------------------------------------------------------------

print("Evaluating the BASELINE RAG on the TEST set...")

baseline_test_eval = evaluate_pipeline_on_goldens(
    answer_function=baseline_answer,
    goldens_subset=test_goldens,
    label="Baseline RAG | Test",
)

print(f"\nBaseline Test Overall Score:  {baseline_test_eval['overall_mean']}")

# Create a dataframe for baseline test evaluation metrics.
baseline_eval_df = pd.DataFrame([
    {
        "Pipeline": "Baseline RAG",
        "Relevance": baseline_test_eval["metric_means"]["Relevance"],
        "Faithfulness": baseline_test_eval["metric_means"]["Faithfulness"],
        "Completeness": baseline_test_eval["metric_means"]["Completeness"],
        "Overall Score": baseline_test_eval["overall_mean"],
    }
])

✨ You're running DeepEval's latest Relevance [GEval] Metric! (using gpt-4o-mini, strict=False, async_mode=True)...

✨ You're running DeepEval's latest Relevance [GEval] Metric! (using gpt-4o-mini, strict=False, async_mode=True)...

✨ You're running DeepEval's latest Relevance [GEval] Metric! (using gpt-4o-mini, strict=False, async_mode=True)...

✨ You're running DeepEval's latest Relevance [GEval] Metric! (using gpt-4o-mini, strict=False, async_mode=True)...

✨ You're running DeepEval's latest Relevance [GEval] Metric! (using gpt-4o-mini, strict=False, async_mode=True)...

In [ ]:
print("\nBaseline Evaluation Summary:")
display(baseline_eval_df)


Baseline Evaluation Summary:


,Pipeline,Relevance,Faithfulness,Completeness,Overall Score
0,Baseline RAG,0.372,0.816,0.812,0.667


# Section 2: Prompt Optimization

## 2.1 Define the prompt template and model callback



***Prompt:***

<font size=3 color="#4682B4"><b>I want to set up DeepEval prompt optimization for a health insurance RAG system. Help me write code to:

1. Create a reusable prompt template using the baseline system prompt and user prompt
2. Replace runtime placeholders so the benchmark question and retrieved context can be injected dynamically
3. Initialize one model for prompt optimization and another for answer generation
4. Use a slightly higher temperature for the optimization model to encourage prompt mutation diversity
5. Create a callback function that:
   - injects the benchmark question and context into the prompt
   - sends the prompt to the LLM
   - returns the generated answer for evaluation

</font>

In [ ]:
# DeepEval prompt optimizer uses a Prompt object as the starting point.

PROMPT_TEMPLATE_V1 = Prompt(
    text_template=(
        BASELINE_SYSTEM_PROMPT
        + "\n\n"
        + BASELINE_USER_PROMPT
            .replace("{question}", "{input}")
            .replace("{context}", "{context}")
    )
)

In [ ]:
# Model used during prompt optimization to suggest prompt mutations.
# NOTE:
# A slightly higher temperature helps encourage prompt mutation diversity.

optim_model = GPTModel(
    model=ANSWER_MODEL,
    temperature=0.5,
    api_key=os.getenv("OPENAI_API_KEY"),
    base_url=os.getenv("OPENAI_BASE_URL"),
)

# LLM used to generate answers during prompt evaluation.

answer_llm = ChatOpenAI(
    model=ANSWER_MODEL,
    temperature=0,
    api_key=os.getenv("OPENAI_API_KEY"),
    base_url=os.getenv("OPENAI_BASE_URL"),
)

In [ ]:
def model_callback(prompt: Prompt, golden: Golden) -> str:
    """
    Called by DeepEval during prompt optimization.

    The function:
    1. Injects the benchmark question and context into the prompt
    2. Sends the final prompt to the LLM
    3. Returns the generated answer
    """

    interpolated = prompt.interpolate(
        input=golden.input,
        context=golden.context[0] if golden.context else "",
    )

    response = answer_llm.invoke([
        SystemMessage(content=(
            "You are a health insurance policy assistant for sales representatives. "
            "Follow the user prompt exactly and answer only from the provided context."
        )),
        HumanMessage(content=interpolated),
    ])

    return response.content

## 2.2 Optimize the prompt with GEPA

***Prompt:***

<font size=3 color="#4682B4"><b>I want to optimize my health insurance RAG prompt automatically using DeepEval’s GEPA algorithm. Help me write code to:

1. Configure retry and timeout settings for DeepEval optimization
2. Initialize the GEPA prompt optimization algorithm
3. Define optimization parameters such as iterations, pareto size, minibatch size, random seed, and tie-breaker strategy
4. Use the benchmark evaluation metrics during optimization
5. Run prompt optimization on the training benchmark dataset
6. Print the original prompt and the optimized prompt for comparison
7. Track and display the execution time for the optimization process

</font>

In [ ]:
# ── GEPA: Genetic-Pareto Prompt Optimisation ──────────────────────────────────
# GEPA iteratively mutates and evaluates prompts to improve performance
# across multiple evaluation metrics.

os.environ["DEEPEVAL_RETRY_MAX_ATTEMPTS"] = "2"
os.environ["DEEPEVAL_RETRY_INITIAL_SECONDS"] = "1.0"
os.environ["DEEPEVAL_RETRY_EXP_BASE"] = "1.0"
os.environ["DEEPEVAL_RETRY_JITTER"] = "1.0"
os.environ["DEEPEVAL_RETRY_CAP_SECONDS"] = "5.0"
os.environ["DEEPEVAL_PER_ATTEMPT_TIMEOUT_SECONDS_OVERRIDE"] = "90"

In [ ]:
start_time = time.time()

# Configure the PromptOptimizer with the GEPA algorithm.
# - iterations: Number of evolutionary steps to perform.
# - pareto_size: Number of top-performing, diverse candidates to track.
# - minibatch_size: Number of golden examples to evaluate per iteration.
# - random_seed: Ensures reproducibility of the mutation/selection process.
# - tie_breaker: Strategy to use when a child prompt has the same score as its parent.

# Configure the PromptOptimizer using the GEPA algorithm.
gepa_optimizer = PromptOptimizer(
    algorithm=GEPA(
        iterations=4,
        pareto_size=2,     # hold 1 frontier candidates
        minibatch_size=4,  # sample 4 samples per evaluation
        random_seed=42,
        tie_breaker=TieBreaker.PREFER_CHILD,  # breaks ties in favour of the mutation
    ),
    model_callback=model_callback,  # The function that executes the prompt and returns output
    metrics=DEEPEVAL_METRICS,       # The list of GEval metrics to optimize for
    optimizer_model=optim_model,    # The LLM that generates the optimized prompt variants
)

# Ensure the algorithm instance specifically uses the provided optimizer model.
gepa_optimizer.algorithm.optimizer_model = optim_model

# Start the optimization process.
# This will iteratively evaluate and mutate the 'PROMPT_TEMPLATE_V1'
# using the provided benchmark dataset.
gepa_prompt = gepa_optimizer.optimize(
    prompt=PROMPT_TEMPLATE_V1,
    goldens=train_goldens,
)

end_time = time.time()

print(f"Execution Time: {end_time - start_time:.4f} seconds")

In [ ]:
print("\nOriginal Prompt:")
print(PROMPT_TEMPLATE_V1.text_template)

print("\nGEPA Optimized Prompt:")
print(gepa_prompt.text_template)


Original Prompt:
You are a health insurance assistant.
Answer the question using the context.

Broker Question:
{input}

Retrieved Context:
{context}

GEPA Optimized Prompt:
You are a health insurance assistant. Answer the question using the context, ensuring to clarify exclusions and specific conditions as needed. 

Broker Question:
{input}

Retrieved Context:
{context}

**Rules:** 1. Clearly state any waiting periods or specific conditions required for coverage. 2. Explicitly mention exclusions such as items listed as 'Not Payable'. 3. Do NOT imply that the company can change policy benefits during the term unless regulatory conditions are met. 4. Ensure that responses regarding domiciliary treatment include the minimum treatment duration and specific conditions that must be satisfied for coverage.


## 2.3 Evaluate the optimized prompt on the benchmark dataset

***Prompt:***

<font size=3 color="#4682B4"><b>I want to evaluate the GEPA-optimized prompt on my health insurance benchmark dataset. Help me write code to:

1. Create a reusable answer-generation function using the optimized prompt
2. Retrieve relevant policy chunks from the vector database
3. Format the retrieved chunks into prompt-ready context
4. Inject the question and context into the optimized prompt dynamically
5. Generate grounded answers using the optimized prompt
6. Evaluate the optimized prompt only on the benchmark test set
7. Display the final evaluation score and execution time

</font>

In [ ]:
# -------------------------------------------------------------------
# Create a runtime function using the GEPA optimized prompt
# -------------------------------------------------------------------

def optimized_prompt_answer(question: str) -> dict:
    """
    Generate answers using:
    - the GEPA optimized prompt
    - the baseline retrieval pipeline
    """

    # Retrieve relevant chunks from the vector database
    docs = retrieve_docs(question, k=4)

    # Convert retrieved chunks into prompt-ready context
    context = format_docs(docs)

    # Inject runtime values into the optimized prompt
    final_prompt = gepa_prompt.interpolate(
        input=question,
        context=context,
    )

    # Generate the final response
    response = answer_llm.invoke([
        SystemMessage(content=(
            "You are a health insurance policy intelligence assistant for sales representatives. "
            "Answer only from the provided context."
        )),
        HumanMessage(content=final_prompt),
    ])

    return {
        "answer": response.content,
        "context": context,
        "docs": docs,
    }

In [ ]:
# -------------------------------------------------------------------
# Evaluate the GEPA optimized prompt only on the TEST set
# -------------------------------------------------------------------

start_time = time.time()

print("Evaluating the GEPA optimized prompt on the TEST set...")

optimized_test_eval = evaluate_pipeline_on_goldens(
    answer_function=optimized_prompt_answer,
    goldens_subset=test_goldens,
    label="GEPA Optimized Prompt | Test",
)

end_time = time.time()

print(f"\nOptimized Test Overall Score: {optimized_test_eval['overall_mean']}")

print(f"\nExecution Time: {end_time - start_time:.2f} seconds")

✨ You're running DeepEval's latest Relevance [GEval] Metric! (using gpt-4o-mini, strict=False, async_mode=True)...

Output()

## 2.4 Compare optimized prompt with the baseline prompt

***Prompt:***

<font size=3 color="#4682B4"><b>I want to compare the performance of the baseline RAG prompt and the GEPA-optimized prompt for health insurance policy question answering. Help me write code to:

1. Create a comparison dataframe containing:
   - overall score
   - relevance score
   - faithfulness score
   - completeness score
2. Compare the baseline prompt and optimized prompt side-by-side
3. Display the evaluation summary in a clean tabular format

</font>

In [ ]:
# Create a dataframe to compare baseline vs optimized prompt performance

comparison_df = pd.DataFrame([
    {
        "Prompt Version": "Baseline Prompt",

        "Overall Score": baseline_test_eval["overall_mean"],

        "Relevance": baseline_test_eval["metric_means"]["Relevance"],

        "Faithfulness": baseline_test_eval["metric_means"]["Faithfulness"],

        "Completeness": baseline_test_eval["metric_means"]["Completeness"],
    },

    {
        "Prompt Version": "GEPA Optimized Prompt",

        "Overall Score": optimized_test_eval["overall_mean"],

        "Relevance": optimized_test_eval["metric_means"]["Relevance"],

        "Faithfulness": optimized_test_eval["metric_means"]["Faithfulness"],

        "Completeness": optimized_test_eval["metric_means"]["Completeness"],
    },
])

In [ ]:
print("\nBaseline vs Optimized Prompt Comparison:")
display(comparison_df)


Baseline vs Optimized Prompt Comparison:


,Prompt Version,Overall Score,Relevance,Faithfulness,Completeness
0,Baseline Prompt,0.667,0.372,0.816,0.812
1,GEPA Optimized Prompt,0.687,0.384,0.843,0.835


##### **NOTE:**

> If the notebook is executed again, the exact metric values may vary slightly because prompt optimization and LLM-based evaluation are not fully deterministic. However, the overall insights and performance trends usually remain consistent across runs.

### **Observations for Section 2 — DeepEval Prompt Optimization**

- The GEPA-optimized prompt performed slightly better overall than the baseline prompt. The overall score improved from `0.667` to `0.687`, showing a modest but clear gain from prompt optimization.

- The faithfulness score improved from `0.816` to `0.843`. This indicates that the optimized prompt stayed a little more grounded in the retrieved health insurance policy context and reduced unsupported or weakly supported claims.

- The completeness score also improved from `0.812` to `0.835`. This shows that the optimized prompt answered the policy questions a bit more fully and covered more of the expected policy detail.

- The relevance score increased only slightly from `0.372` to `0.384`. This suggests that although the optimized prompt improved grounding and completeness, it still did not closely match the benchmark answers in wording or response structure.

- The overall improvement is small rather than dramatic, which indicates that the baseline prompt was already reasonably strong for this dataset and that GEPA mainly helped refine answer quality rather than transform it.

- The gap between faithfulness and relevance remains noticeable in both prompts. This means the system is better at staying grounded in the retrieved policy context than at matching the exact expected answer style or phrasing.

- Overall, the GEPA optimization improved the reliability and completeness of the health insurance RAG system, but there is still room to improve answer relevance and benchmark alignment for more precise policy Q&A.

# Section 3: Tuned RAG Pipeline


## 3.1 Define the RAG configurations

***Prompt:***

<font size=3 color="#4682B4"><b>I want to experiment with multiple RAG runtime configurations for a health insurance policy intelligence system. Help me write code to:

1. Define multiple RAG configurations with different retrieval and generation settings
2. Tune parameters such as:
   - number of retrieved chunks (`k`)
   - temperature
   - top_p
   - max_tokens
3. Create configurations for:
   - precision-focused retrieval
   - balanced retrieval
   - context-heavy retrieval
4. Add a short rationale explaining the purpose of each configuration
5. Print all configurations for verification

</font>

In [ ]:
# Each configuration uses different retrieval and generation settings.

# We tune only runtime parameters:
# - k: number of retrieved chunks
# - temperature: randomness of generation
# - top_p: nucleus sampling
# - max_tokens: response length

RAG_CONFIGURATIONS = [
    {
        "name": "Config_1_Precision_Focused",
        "k": 3,
        "temperature": 0.0,
        "top_p": 0.9,
        "max_tokens": 400,
        "rationale": (
            "Focused retrieval with fewer chunks and low randomness. "
            "Best for concise, grounded policy answers."
        ),
    },
    {
        "name": "Config_2_Balanced",
        "k": 4,
        "temperature": 0.2,
        "top_p": 0.95,
        "max_tokens": 600,
        "rationale": (
            "Balanced setup for general sales and policy questions. "
            "Good mix of grounding, coverage, and answer length."
        ),
    },
    {
        "name": "Config_3_Context_Heavy",
        "k": 6,
        "temperature": 0.3,
        "top_p": 1.0,
        "max_tokens": 800,
        "rationale": (
            "Broader retrieval for cross-document reasoning and synthesis-heavy questions. "
            "Useful when the answer needs more context."
        ),
    },
]

print("Defined RAG configurations:\n")
for config in RAG_CONFIGURATIONS:
    print(f"{config['name']}")
    print(f"Rationale: {config['rationale']}\n")

Defined RAG configurations:

Config_1_Precision_Focused
Rationale: Focused retrieval with fewer chunks and low randomness. Best for concise, grounded policy answers.

Config_2_Balanced
Rationale: Balanced setup for general sales and policy questions. Good mix of grounding, coverage, and answer length.

Config_3_Context_Heavy
Rationale: Broader retrieval for cross-document reasoning and synthesis-heavy questions. Useful when the answer needs more context.



## 3.2 Create a runtime answering function

***Prompt:***

<font size=3 color="#4682B4"><b>I want to create a reusable answer-generation function that supports different RAG runtime configurations. Help me write code to:

1. Retrieve context dynamically using the configuration’s `k` value
2. Format retrieved chunks into prompt-ready context
3. Inject the question and context into the GEPA-optimized prompt
4. Create a ChatOpenAI model dynamically using runtime generation parameters
5. Generate grounded health insurance policy answers using the optimized prompt
6. Return the generated answer, retrieved context, and source documents

</font>

In [ ]:
def answer_with_rag_config(question: str, config: dict) -> dict:
    """
    Answer a question using:
    - the fixed Chroma vector store,
    - the GEPA-optimized prompt,
    - and the runtime configuration settings.
    """

    # Retrieve the most relevant chunks using the current k value.
    docs = vectorstore.similarity_search(question, k=config["k"])

    # Turn the retrieved chunks into a prompt-ready context block.
    context = format_docs(docs)

    # Inject the question and retrieved context into the optimized prompt.
    final_prompt = gepa_prompt.interpolate(
        input=question,
        context=context,
    )

    # Create the answer model using the current config settings.
    # top_p and max_tokens are passed through model_kwargs.
    config_llm = ChatOpenAI(
        model=ANSWER_MODEL,
        temperature=config["temperature"],
        model_kwargs={
            "top_p": config["top_p"],
            "max_tokens": config["max_tokens"],
        },
        api_key=os.getenv("OPENAI_API_KEY"),
        base_url=os.getenv("OPENAI_API_BASE"),
    )

    # Ask the model to generate the final answer.
    response = config_llm.invoke([
        SystemMessage(content=(
            "You are a health insurance policy intelligence assistant for sales representatives. "
            "Answer only from the provided context and stay grounded."
        )),
        HumanMessage(content=final_prompt),
    ])

    return {
        "answer": response.content,
        "context": context,
        "docs": docs,
    }

## 3.3 Evaluate one configuration on a benchmark set

***Prompt:***

<font size=3 color="#4682B4"><b>I want to evaluate one RAG configuration on a benchmark dataset for health insurance policy QA. Help me write code to:

1. Create a reusable evaluation function for a single RAG configuration
2. Generate answers using the selected runtime configuration
3. Build DeepEval test cases for each benchmark example
4. Measure relevance, faithfulness, and completeness scores
5. Store metric scores across all benchmark examples
6. Compute average metric values and an overall evaluation score
7. Return the evaluation results along with the configuration details

</font>

In [ ]:
def evaluate_rag_configuration(config: dict, goldens_subset, label: str) -> dict:
    """
    Evaluate one RAG configuration on a given benchmark subset.

    Returns:
    - metric-wise average scores
    - overall average score
    """

    print("\n" + "=" * 90)
    print(f"Evaluating: {config['name']} | {label}")
    print("=" * 90)

    metric_scores = {
        metric.name: []
        for metric in DEEPEVAL_METRICS
    }

    for idx, golden in enumerate(goldens_subset, start=1):
        print(f"\nEvaluating Example {idx}/{len(goldens_subset)}")

        # Generate answer using the current config
        result = answer_with_rag_config(
            question=golden.input,
            config=config,
        )

        # Build the DeepEval test case
        test_case = LLMTestCase(
            input=golden.input,
            actual_output=result["answer"],
            expected_output=golden.expected_output,
            context=[result["context"]],
        )

        # Measure all metrics
        for metric in DEEPEVAL_METRICS:
            try:
                metric.measure(test_case)

                if metric.score is not None:
                    metric_scores[metric.name].append(metric.score)

                print(f"{metric.name}: {metric.score}")

            except Exception as e:
                print(f"{metric.name} failed: {e}")

    # Compute average scores for each metric
    metric_means = {
        name: round(sum(scores) / len(scores), 3) if scores else 0.0
        for name, scores in metric_scores.items()
    }

    # Overall score is the mean of all metric means
    overall_score = round(
        sum(metric_means.values()) / len(metric_means),
        3
    ) if metric_means else 0.0

    print("\nMetric Means:")
    print(metric_means)
    print(f"Overall Score: {overall_score}")

    return {
        "configuration": config["name"],
        "metric_means": metric_means,
        "overall_score": overall_score,
        "rationale": config["rationale"],
    }

## 3.4 Run all configurations on the test set

***Prompt:***

<font size=3 color="#4682B4"><b>I want to evaluate all defined RAG configurations on the benchmark test set. Help me write code to:

1. Run every RAG configuration sequentially on the test dataset
2. Store the evaluation results for each configuration
3. Measure the total execution time for the tuning process
4. Print confirmation once all configuration evaluations are completed

</font>

In [ ]:
configuration_results = []

start_time = time.time()

for config in RAG_CONFIGURATIONS:
    result = evaluate_rag_configuration(
        config=config,
        goldens_subset=test_goldens,
        label="Test Set",
    )
    configuration_results.append(result)

end_time = time.time()

print("\nCompleted evaluation on the test set.")
print(f"Execution Time: {end_time - start_time:.2f} seconds")

✨ You're running DeepEval's latest Completeness [GEval] Metric! (using gpt-4o-mini, strict=False, async_mode=True…

Completeness: 0.7880797073284305

Evaluating Example 2/6


Output()

## 3.5 Compare configurations

***Prompt:***

<font size=3 color="#4682B4"><b>I want to compare all evaluated approaches for my health insurance RAG system. Help me write code to:

1. Combine:
   - baseline prompt results
   - GEPA optimized prompt results
   - all tuned RAG configuration results
2. Create a comparison dataframe containing:
   - relevance
   - faithfulness
   - completeness
   - overall score
   - rationale
3. Sort all approaches by overall score
4. Display the final comparison table for analysis

</font>

In [ ]:
# Add baseline and optimized prompt results so all five approaches are compared together.
all_results = [
    {
        "Approach": "Baseline Prompt",
        "Relevance": baseline_test_eval["metric_means"].get("Relevance", 0.0),
        "Faithfulness": baseline_test_eval["metric_means"].get("Faithfulness", 0.0),
        "Completeness": baseline_test_eval["metric_means"].get("Completeness", 0.0),
        "Overall Score": baseline_test_eval["overall_mean"],
        "Rationale": "Baseline prompt from Section 1.",
    },
    {
        "Approach": "GEPA Optimized Prompt",
        "Relevance": optimized_test_eval["metric_means"].get("Relevance", 0.0),
        "Faithfulness": optimized_test_eval["metric_means"].get("Faithfulness", 0.0),
        "Completeness": optimized_test_eval["metric_means"].get("Completeness", 0.0),
        "Overall Score": optimized_test_eval["overall_mean"],
        "Rationale": "GEPA-optimized prompt from Section 2.",
    },
]

# Add the three runtime RAG configurations.
for result in configuration_results:
    all_results.append({
        "Approach": result["configuration"],
        "Relevance": result["metric_means"].get("Relevance", 0.0),
        "Faithfulness": result["metric_means"].get("Faithfulness", 0.0),
        "Completeness": result["metric_means"].get("Completeness", 0.0),
        "Overall Score": result["overall_score"],
        "Rationale": result["rationale"],
    })

results_df = pd.DataFrame(all_results)

results_df = results_df.sort_values(
    by="Overall Score",
    ascending=False
).reset_index(drop=True)

print("\nRAG Configuration Comparison on Test Set:")
display(results_df)


RAG Configuration Comparison on Test Set:


,Approach,Relevance,Faithfulness,Completeness,Overall Score,Rationale
0,GEPA Optimized Prompt,0.384,0.843,0.835,0.687,GEPA-optimized prompt from Section 2.
1,Config_3_Context_Heavy,0.411,0.822,0.829,0.687,Broader retrieval for cross-document reasoning...
2,Config_2_Balanced,0.387,0.822,0.827,0.679,Balanced setup for general sales and policy qu...
3,Config_1_Precision_Focused,0.391,0.819,0.805,0.672,Focused retrieval with fewer chunks and low ra...
4,Baseline Prompt,0.372,0.816,0.812,0.667,Baseline prompt from Section 1.


## 3.6 Select the best overall approach

***Prompt:***

<font size=3 color="#4682B4"><b>I want to automatically select the best-performing RAG approach based on benchmark evaluation results. Help me write code to:

1. Identify the highest-scoring approach from the comparison dataframe
2. Extract the corresponding configuration or prompt details
3. Store the selected approach in a reusable variable
4. Print the selected approach and its details in a readable format

</font>

In [ ]:
best_approach_name = results_df.iloc[0]["Approach"]

BEST_FINAL_APPROACH = next(
    item for item in all_results
    if item["Approach"] == best_approach_name
)

print("\nBest Approach Selected:")
print(BEST_FINAL_APPROACH["Approach"])
print("\nDetails:")
print(json.dumps(BEST_FINAL_APPROACH, indent=2))


Best Approach Selected:
GEPA Optimized Prompt

Details:
{
  "Approach": "GEPA Optimized Prompt",
  "Relevance": 0.384,
  "Faithfulness": 0.843,
  "Completeness": 0.835,
  "Overall Score": 0.687,
  "Rationale": "GEPA-optimized prompt from Section 2."
}


## 3.7 Save the tuning results

***Prompt:***

<font size=3 color="#4682B4"><b>I want to save the RAG tuning and evaluation results for future analysis. Help me write code to:

1. Save the full configuration comparison dataframe as a CSV file
2. Save the best-performing approach and all evaluation results as a JSON file
3. Store the files inside the artifacts directory
4. Print the saved file locations for verification

</font>

In [ ]:
results_output_path = ARTIFACT_DIR / "rag_configuration_results.csv"
results_df.to_csv(results_output_path, index=False)

best_approach_path = ARTIFACT_DIR / "best_final_approach.json"

with open(best_approach_path, "w", encoding="utf-8") as f:
    json.dump(
        {
            "best_approach": BEST_FINAL_APPROACH,
            "all_test_results": all_results,
        },
        f,
        indent=2,
        ensure_ascii=False,
    )

print(f"\nSaved configuration comparison results to: {results_output_path}")
print(f"Saved best approach to: {best_approach_path}")


Saved configuration comparison results to: artifacts/rag_configuration_results.csv
Saved best approach to: artifacts/best_final_approach.json


##### **NOTE:**

> The evaluation scores may vary slightly across notebook executions because LLM generation and DeepEval scoring are probabilistic. However, the ranking and overall behavior of the configurations generally remain stable.

### **Observations for Section 3 — RAG Configuration Tuning and Evaluation**

- Among all the approaches tested, both the **GEPA Optimized Prompt** and **Config_3_Context_Heavy** achieved the highest overall score (`0.687`). This indicates that both prompt optimization and broader retrieval contributed positively to the quality of the health insurance policy answering system.

- **Config_3_Context_Heavy** achieved the highest relevance score (`0.411`) among all approaches. This suggests that retrieving more policy context helped the model better address multi-document insurance questions involving exclusions, waiting periods, claims, and coverage conditions.

- The **GEPA Optimized Prompt** achieved the strongest faithfulness score (`0.843`) and the highest completeness score (`0.835`) overall. This indicates that prompt optimization improved groundedness and helped the model generate more complete policy responses from the retrieved context.

- **Config_2_Balanced** performed consistently across all three metrics and achieved an overall score of `0.679`. This shows that moderate retrieval depth and controlled generation settings provided stable performance, even though they did not outperform the broader retrieval strategy.

- **Config_1_Precision_Focused** achieved slightly lower completeness (`0.805`) compared to the other tuned configurations. Retrieving fewer chunks reduced retrieval noise, but it also limited the amount of policy detail available for complex or cross-document insurance questions.

- The **Baseline Prompt** produced the lowest overall score (`0.667`) among all approaches. While it remained reasonably grounded, it was less complete and less effective at handling detailed policy reasoning compared to the optimized and tuned approaches.

- One important pattern across the results is that health insurance policy questions often require combining information from multiple documents such as policy brochures, exclusions, terms and conditions, and claim guides. Broader retrieval therefore improved answer quality for many benchmark scenarios.

- The differences between the overall scores were relatively small, which suggests that retrieval quality, prompt grounding, and answer completeness were already reasonably stable across configurations. However, the optimized prompt and context-heavy retrieval strategy still produced the strongest overall performance.

# Section 4: Final RAG Pipeline

## 4.1 Define the final inference helper

***Prompt:***

<font size=3 color="#4682B4"><b>I want to build a final inference function for my health insurance RAG system. Help me write code to:

1. Automatically use the best-performing approach selected during evaluation
2. Support:
   - baseline prompt inference
   - GEPA optimized prompt inference
   - tuned RAG configuration inference
3. Retrieve relevant policy chunks from ChromaDB
4. Generate grounded answers using the selected prompt and runtime settings
5. Return the final answer along with retrieved documents and metadata

</font>

In [ ]:
def answer_final_question(question: str) -> dict:
    """
    Generate a final answer using the best-performing approach selected in Section 3.
    The best approach may be:
    - Baseline Prompt
    - GEPA Optimized Prompt
    - One of the three RAG configurations
    """

    # Read the best approach selected in Section 3.
    best_approach_name = BEST_FINAL_APPROACH["Approach"]

    # -------------------------------------------------------------------
    # Case 1: Baseline Prompt
    # -------------------------------------------------------------------
    if best_approach_name == "Baseline Prompt":

        # Retrieve the most relevant chunks using the default retrieval setting.
        docs = retrieve_docs(question, k=4)

        # Convert retrieved chunks into prompt-ready context.
        context = format_docs(docs)

        # Build the baseline user prompt.
        user_prompt = BASELINE_USER_PROMPT.format(
            question=question,
            context=context,
        )

        # Generate the answer using the baseline LLM.
        response = llm.invoke([
            SystemMessage(content=BASELINE_SYSTEM_PROMPT),
            HumanMessage(content=user_prompt),
        ])

        return {
            "question": question,
            "answer": response.content,
            "docs": docs,
            "context": context,
            "approach": best_approach_name,
        }

    # -------------------------------------------------------------------
    # Case 2: GEPA Optimized Prompt
    # -------------------------------------------------------------------
    if best_approach_name == "GEPA Optimized Prompt":

        # Retrieve the most relevant chunks using the default retrieval setting.
        docs = retrieve_docs(question, k=4)

        # Convert retrieved chunks into prompt-ready context.
        context = format_docs(docs)

        # Inject the question and context into the optimized prompt.
        final_prompt = gepa_prompt.interpolate(
            input=question,
            context=context,
        )

        # Generate the final answer with the optimized prompt.
        response = answer_llm.invoke([
            SystemMessage(content=(
                "You are an AI-powered health insurance policy intelligence assistant for sales representatives. "
                "Answer only from the provided context. "
                "Be concise, grounded, and evidence-based."
            )),
            HumanMessage(content=final_prompt),
        ])

        return {
            "question": question,
            "answer": response.content,
            "docs": docs,
            "context": context,
            "approach": best_approach_name,
        }

    # -------------------------------------------------------------------
    # Case 3: One of the RAG configurations
    # -------------------------------------------------------------------
    config = next(
        cfg for cfg in RAG_CONFIGURATIONS
        if cfg["name"] == best_approach_name
    )

    # Retrieve the most relevant chunks using the tuned k value.
    docs = vectorstore.similarity_search(question, k=config["k"])

    # Convert retrieved chunks into prompt-ready context.
    context = format_docs(docs)

    # Inject the question and context into the optimized prompt.
    final_prompt = gepa_prompt.interpolate(
        input=question,
        context=context,
    )

    # Create the LLM using the selected runtime settings.
    final_llm = ChatOpenAI(
        model=ANSWER_MODEL,
        temperature=config["temperature"],
        model_kwargs={
            "top_p": config["top_p"],
            "max_tokens": config["max_tokens"],
        },
        api_key=os.getenv("OPENAI_API_KEY"),
        base_url=os.getenv("OPENAI_API_BASE"),
    )

    # Generate the final grounded answer.
    response = final_llm.invoke([
        SystemMessage(content=(
            "You are an AI-powered health insurance policy intelligence assistant supporting sales representatives. "
            "Answer only from the retrieved context. "
            "Be concise, grounded, and evidence-based."
        )),
        HumanMessage(content=final_prompt),
    ])

    return {
        "question": question,
        "answer": response.content,
        "docs": docs,
        "context": context,
        "approach": best_approach_name,
        "configuration": config["name"],
    }

## 4.2 Define the final test cases

***Prompt:***

<font size=3 color="#4682B4"><b>I want to create realistic test questions for evaluating a health insurance policy intelligence system. Help me write code to:

1. Create multiple customer-style insurance questions
2. Include scenarios related to:
   - waiting periods
   - recharge benefits
   - room-rent limits
   - reimbursement claims
   - maternity and domiciliary coverage
3. Store all questions as structured test cases for final inference evaluation

</font>

In [ ]:
final_test_cases = [
    {
        "id": 1,
        "question": (
            "If a customer with diabetes buys the policy today, "
            "when will diabetes-related hospitalization become covered, "
            "and does the Instant Cover rider change the waiting period?"
        )
    },
    {
        "id": 2,
        "question": (
            "A family has exhausted their entire base sum insured on one member’s surgery. "
            "If another family member gets hospitalized later in the same year, "
            "how does the Unlimited Automatic Recharge feature work?"
        )
    },
    {
        "id": 3,
        "question": (
            "What happens if a customer chooses a room category higher than "
            "their eligible room type during hospitalization? "
            "Which expenses are reduced under pro-rata deduction, "
            "and which expenses remain fully payable?"
        )
    },
    {
        "id": 4,
        "question": (
            "If a customer goes to a non-network hospital during an emergency, "
            "what reimbursement rules, timelines, and co-payment conditions apply?"
        )
    },
    {
        "id": 5,
        "question": (
            "Compare how the policy handles maternity coverage, "
            "day care procedures like cataract surgery, "
            "and domiciliary hospitalization treatment at home."
        )
    },
]


## 4.3 Run the final test cases

***Prompt:***

<font size=3 color="#4682B4"><b>I want to run the final health insurance policy test cases through the optimized RAG pipeline. Help me write code to:

1. Execute all final test questions one by one
2. Generate answers using the final inference function
3. Display:
   - question
   - generated answer
   - selected approach
   - source files used
4. Save all final inference outputs into a structured results list

</font>

In [ ]:
## 4.3 Run the final test cases

final_results = []

for item in final_test_cases:
    print("\n" + "=" * 100)
    print(f"Test Case {item['id']}")
    print("=" * 100)

    result = answer_final_question(item["question"])

    # Save a short source summary for easier review.
    source_files = sorted(
        {
            doc.metadata.get("source_file", "unknown")
            for doc in result["docs"]
        }
    )

    print("\nQuestion:")
    print(result["question"])

    print("\nAnswer:")
    print(result["answer"])

    print("\nApproach used:")
    print(result["approach"])

    print("\nSource files used:")
    print(", ".join(source_files))

    final_results.append({
        "id": item["id"],
        "question": result["question"],
        "answer": result["answer"],
        "approach": result["approach"],
        "configuration": result.get("configuration", ""),
        "source_files": ", ".join(source_files),
    })


Test Case 1

Question:
If a customer with diabetes buys the policy today, when will diabetes-related hospitalization become covered, and does the Instant Cover rider change the waiting period?

Answer:
If a customer with diabetes purchases the policy today, diabetes-related hospitalization will be covered starting from the 31st day of the policy start date, following an initial 30-day waiting period. 

The Instant Cover rider, if opted, waives the applicable waiting period for diabetes, allowing coverage to commence from the 31st day as well. However, it's important to note that the customer can choose either the Instant Cover or the Pre-Existing Disease (PED) Wait Period Modification, but not both. 

Make sure to clarify that any claims related to diabetes will only be payable after this waiting period, and ensure the customer understands the implications of choosing one rider over the other.

Approach used:
GEPA Optimized Prompt

Source files used:
Health_Insurance_Policy_TandC.pdf


# Conclusion

## Actionable Insights:

1. The proof-of-concept demonstrates that MeridianHealth can significantly improve the consistency and reliability of policy guidance provided during live sales conversations. By combining RAG with automated prompt optimization, the system was able to generate more grounded and complete responses while reducing the likelihood of unsupported policy statements.

2. The benchmark-driven evaluation framework provides MeridianHealth with a measurable and repeatable quality-control mechanism before deploying AI-assisted policy guidance to the field. This creates a stronger compliance and governance foundation for scaling AI-powered sales enablement in a regulated insurance environment.

3. The experiments confirmed that health insurance policy servicing is inherently a multi-document reasoning problem. Questions related to exclusions, waiting periods, reimbursement conditions, and recharge benefits often require information from several documents simultaneously, making broader retrieval strategies operationally more effective.

4. The comparison between baseline and optimized workflows highlights that retrieval quality and prompt quality directly influence customer-facing answer reliability. Optimized prompts improved groundedness and completeness, while tuned retrieval settings improved context coverage and clause-level reasoning accuracy.

5. The modular architecture built using LangChain, ChromaDB, OpenAI models, and DeepEval creates a scalable foundation for MeridianHealth’s future AI initiatives. The same framework can later support broker enablement, underwriting support, policy servicing, claims assistance, compliance monitoring, and internal knowledge-management workflows.

6. Based on the answers generated and the evaluation results, we observe that many insurance-policy answers can still be correct and grounded even when they do not exactly match the benchmark answer wording. Variations in explanation style, clause interpretation, and amount of detail are the likely reasons for the reduced relevance scores, despite the response being factually accurate and complete.

## Recommendations:

1. Expand the benchmark dataset with more business-critical scenarios such as state-specific policy rules, portability conditions, underwriting exceptions, rider combinations, claim disputes, and escalation cases. This will improve evaluation realism and ensure the system is tested against the kinds of questions sales representatives face in production.

2. Introduce more advanced retrieval strategies such as metadata-aware filtering, hybrid semantic-keyword retrieval, reranking models, and adaptive chunking. These improvements can help MeridianHealth surface highly relevant clauses faster and improve answer precision for complex policy and claims queries.

3. Standardize benchmark-answer formats and clause-level expected outputs to better reflect insurance-response requirements. This will improve evaluation consistency and reduce cases where generated answers are operationally correct but receive lower relevance scores due to differences in phrasing or response structure.

4. Operationalize the solution as a production-ready AI assistance platform by separating ingestion, retrieval, evaluation, optimization, and inference into reusable APIs and services. This will improve scalability, maintainability, auditability, and deployment readiness across MeridianHealth’s direct sales and broker ecosystem.

<font size=6 color='#4682B4'>Power Ahead</font>
___